### ISSUE TITLE
TotalSegmentator (104 organs) by MONAI team | MONAI Label | Bundle

### ISSUE THREAD

**Issue by tangy5** [Project-MONAI/MONAILabel#1269]
3D Slicer plugin to inference (Andras) CT imaging gaining momentum. 

We will support TotalSgementator model trained by MONAI team instead, 104 organs based on SegResNet.

Train and Benchmarking on A100 GPUs, get performance, time, GPU utility, hardware requirement. 
Best performed model

We will create a bundle for complete pipeline: inference, train, multi-gpu train.

Andres has created a pipeline with Label, we will continue to work on the bundle, benchmarking, A100 Training...

**Comment by pieper** [comment]
Hi @tangy5 - this is a great idea, thanks for working on this.  I have tried the MONAI Label model trained on the TotalSegmentator dataset and it works well, in some cases faster than TotalSegmentator on the same data.  The only problem I had is that the GPU and CPU memory resources can be demanding for the MONAI Label version, whereas the nnUNET-based TotalSegmentator reliably works on with about 11 GB of GPU memory.

Do you think the MONAI Label inference can be tweaked to work with fewer resources?

**Comment by lassoan** [comment]
Note that to reproduce the wide success of the current nnunet-based implementation of TotalSegmentator, it is important to not just achieve similar robustness and accuracy, but also inference speed:
- on CPU: low-resolution (3mm isotropic) segmentation under 2 minutes, full resolution (1.5mm isotropic) segmentation in 30-40 minutes
- on GPU (minimum 7GB memory): low resolution under 30 seconds, full resolution under 1 minute

TotalSegmentator has probably shifted user expectations for CT segmentation in general. Now they expect that segmentation is feasible on the CPU and should just take minutes if they don't need high-resolution results.

**Comment by lassoan** [comment]
It may be also useful for you to know that Jakob Wasserthal said he was open to try any other segmentation model that someone recommends him, but it has to work similarly to the current one in term of performance (robustness, accuracy, speed) and it should not require any manual setting of training parameters, because he wants to use his time for improving the training data set, and not with parameter tuning.

Therefore, you have an opportunity here: if you can provide him a MONAI-based solution that fulfills these requirements (similar in most aspects and maybe significantly better in some) then he may switch to use MONAI. He is still actively adding new structures, considering adding new imaging modalities, etc. so even if MONAI cannot do significantly better than the current solution for the current dataset, it may be able to prove itself as the size and complexity of the data set grows.

**Comment by deepakri201** [comment]
I just wanted to mention that I was able to create the MONAI bundle (for inference) for the full CT segmentation. There were a few transforms I had to modify (for instance I had to remove the Orientationd transform) but here are some preliminary results on an example from the spleen medical decathalon dataset:  

![02_01_23_full_ct_segmentation_success](https://user-images.githubusercontent.com/59979551/216035457-a2fe14b5-91f6-4231-b7e2-0a4a1ba1e507.JPG)

**Comment by tangy5** [comment]
> Hi @tangy5 - this is a great idea, thanks for working on this. I have tried the MONAI Label model trained on the TotalSegmentator dataset and it works well, in some cases faster than TotalSegmentator on the same data. The only problem I had is that the GPU and CPU memory resources can be demanding for the MONAI Label version, whereas the nnUNET-based TotalSegmentator reliably works on with about 11 GB of GPU memory.
> 
> Do you think the MONAI Label inference can be tweaked to work with fewer resources?

MONAI Label version now generate 104 channels at the same time (single model), instead of separate models. We are discussing this within the team, and see there are better options. If we separate the 104 into 5 models as mentioned in the paper, we can safely decrease the resources demand. Now the single model can use SlidingWindowInference with CPU device, the speed is not bad.

**Comment by tangy5** [comment]
> It may be also useful for you to know that Jakob Wasserthal said he was open to try any other segmentation model that someone recommends him, but it has to work similarly to the current one in term of performance (robustness, accuracy, speed) and it should not require any manual setting of training parameters, because he wants to use his time for improving the training data set, and not with parameter tuning.
> 
> Therefore, you have an opportunity here: if you can provide him a MONAI-based solution that fulfills these requirements (similar in most aspects and maybe significantly better in some) then he may switch to use MONAI. He is still actively adding new structures, considering adding new imaging modalities, etc. so even if MONAI cannot do significantly better than the current solution for the current dataset, it may be able to prove itself as the size and complexity of the data set grows.

Thanks, we will discuss this within the MONAI team.

**Comment by kbressem** [comment]
I was also able to recreate the 3mm model as a single model achieving nearly the same performance as TotalSegmentator fast (-1%). However, with the large model, I frequently get OOM errors, not during the sliding window inference but when `AsDiscrete`is applied, even on 80GB A100 GPUs with batch size 1.

**Comment by lassoan** [comment]
Thanks for the update @kbressem, this sounds very promising. Does the inference works as fast on the CPU as well, or you only tested on the GPU?

To avoid requiring large amount of memory, TotalSegmentator splits the segments into 4-5 groups for the full-resolution segmentation. In addition to reducing memory needs, this also allows faster inference (if somebody needs only certain segments then it may be enough to just run only 1 of the 5 models).

**Comment by kbressem** [comment]
I now did some testing on my 2019 MB Pro. Inference time heavily depends on the overlap in the sliding window inferer. With setting the overlap to 0.25 and the batch size to 4, it took about 37s to process a single image of 512 x 512 x 236 px. 17s of this were pure inference time, the rest was initializing the bundle, loading the model, pre/postprocessing and saving. 
Increase in RAM was about 3GB, but this is just am estimate from `htop`. 

Output looks like this: 

<img width="1441" alt="image" src="https://user-images.githubusercontent.com/37253540/217333812-df1a97ff-b0ef-47fe-b91e-1ea2b5352f5d.png">

**Comment by tangy5** [comment]
> I now did some testing on my 2019 MB Pro (2,4 GHz Quad-Core Intel Core i5). Inference time heavily depends on the overlap in the sliding window inferer. With setting the overlap to 0.25 and the batch size to 4, it took about 37s to process a single image of 512 x 512 x 236 px. 17s of this were pure inference time, the rest was initializing the bundle, loading the model, pre/postprocessing and saving. Increase in RAM was about 3GB, but this is just am estimate from `htop`.
> 
> Output looks like this:
> 
> <img alt="image" width="1441" src="https://user-images.githubusercontent.com/37253540/217333812-df1a97ff-b0ef-47fe-b91e-1ea2b5352f5d.png">

This is the inference for the 3mm model right?

**Comment by kbressem** [comment]
Yes. Only 3 mm. The 1.5 mm I tried larger models but had OOM errors during validation.

**Comment by tangy5** [comment]
@kbressem , I guess the problem could be the post transform. AsDiscrete and Invert transform matters here. 
BTW, the MONAI Team is also creating a bundle that can be usefull for the community. 
We have a draft version, most experiments and tests are done. Now we are creating docs and some benchmarking. 

If you'd like to try ahead of release, here is the bundle: https://github.com/tangy5/model-zoo/tree/segresnet_totalSegmentator.

I will provide some numbers and screenshot in another note. Thank you!

**Comment by SachidanandAlle** [comment]
You can make this as config option and pass them through client request (3D slicer options tab) to allow user to run either version of the model. Also enable sw_batch size etc.. as config

**Comment by kbressem** [comment]
I attempted to evaluate the 1.5 mm model on the TotalSegmentator test dataset but faced memory issues (OOM) in larger scans (e.g. angiography exams in the TotalSegmentator dataset), even when utilizing A100 GPUs. The error occurred during post-processing specifically in the `AsDiscreted` step. 
Despite efforts to reduce the memory needs by reducing the precision of the prediction and label to `float16` and `uint8` and removing all other elements in the data dictionary, the issue persisted. 

Switching to a CPU could be an option, but the process is very slow, and it's unlikely that most normal workstations would have enough memory to support it. 

I also tried running `AsDiscreted` on patches of the tensor, but this did not solve the issue as well. I guess, as the tensor is already present in memory, splitting it into smaller parts does not reduce the memory requirement. 

At the moment, the only solution I see is dividing the input image into smaller parts, running inference separately and then merging again. Or are there any other ways to reduce the memory requirement of the large model.

**Comment by tangy5** [comment]
Thought about the options that users can control in the Slicer UI for the totalSegmentator inference. 

See the screenshot, users can choose whether to use lowres model (3.0mm iso) or highres mode (1.5mm iso).
For speed, users can also set sw_overlap, smaller can lead to faster inference speed.  

Any comments?

![options](https://user-images.githubusercontent.com/58751975/218186297-55725012-9c23-4062-8725-f19519c4b1a8.png)

**Comment by tangy5** [comment]
HI team, this is done. We will make some demos, tutorials, docs for this model trained with TotalSegmentator data.

Bundle side:https://github.com/Project-MONAI/model-zoo/tree/dev/models/wholeBody_ct_segmentation
Label configs: https://github.com/Project-MONAI/MONAILabel/pull/1302

Feel free to reopen it if there are further questions.
